### PR TITLE
ITF Fake Peers: behaviours & internal communication capabilities

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

This makes ITF Fake Peers apt for testing basic internal network communication between Iroha peers. Fake Peers get all needed network stubs and handles to simulate network environment in ITF tests. Also Fake Peers can be assigned standard or custom behaviours to reuse some common automation patterns.

### Benefits

Extended network testing facilities in ITF.

### Possible Drawbacks 

A this point we need to be extra careful with synchronization when using Fake Peers in tests, as they run asynchronously. Careless usage may lead to randomly failing tests.

### Usage Examples or Tests *[optional]*

Please see `fake_peer_example_test`.

### CODE REVIEW

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      virtual ~Behaviour();
+
+      /// Enable the behaviour for the given peer
+      void adopt(const std::shared_ptr<FakePeer> &fake_peer);
+
+      /// Disable the behaviour
+      void absolve();
+
+      /// This method gets subscribed on Fake Peer's MST messages.
+      virtual void processMstMessage(MstMessagePtr message) = 0;
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/behaviour/behaviour.hpp]: probably default implementation as "do nothing" for these abstract methods might be ok

File: test/framework/integration_framework/fake_peer/block_storage.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+#include <boost/algorithm/string/join.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/algorithm/max_element.hpp>
+#include "backend/protobuf/block.hpp"
+#include "framework/integration_framework/fake_peer/fake_peer.hpp"
+
+/// Emplace to a map.
+/// @return true if overwritten.
+template <typename MapType, typename KeyType, typename ValueType>
+static bool emplaceCheckingOverwrite(MapType &map,
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/block_storage.cpp:20]: wondering for moving to c++17 :( https://en.cppreference.com/w/cpp/container/map/insert_or_assign

File: test/framework/integration_framework/fake_peer/block_storage.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      return found->second;
+    }
+
+    BlockStorage::BlockPtr BlockStorage::getTopBlock() const {
+      if (blocks_by_height_.empty()) {
+        getLogger()->info(
+            "Requested top block, but the block storage is empty.");
+        return {};
+      }
+      return blocks_by_height_.at(*boost::range::max_element(
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/block_storage.cpp:85]: can be operator[]

File: test/framework/integration_framework/fake_peer/block_storage.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        getLogger()->warn(
+            "Requested removal of a using fake peer "
+            "without previously registering it.");
+        return;
+      }
+      using_peers_.erase(found);
+    }
+
+    logger::Logger BlockStorage::getLogger() const {
+      std::vector<std::string> using_peers_addresses;
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/block_storage.cpp]: Не очень имя, в первый раз перепутал с `using_peers_` и долго не мог понять у кого вызывается lock. Мб лучше `addresses_str`?

File: test/framework/integration_framework/fake_peer/block_storage.hpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      void claimNotUsingPeer(const std::shared_ptr<FakePeer> &peer);
+
+     private:
+      logger::Logger getLogger() const;
+
+      std::unordered_map<HeightType, BlockPtr> blocks_by_height_;
+      std::unordered_map<HashType, BlockPtr, HashType::Hasher> blocks_by_hash_;
+
+      /// The collection of peers claiming to use this block storage.
+      // I am using vector (although a kind of set would fit better) because
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/block_storage.hpp]: I am using vector -> Vector is used

File: test/framework/integration_framework/fake_peer/block_storage.hpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      // weak pointers seem unsuitable for any comparison, as the shared pointer
+      // they point to may change.
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/block_storage.hpp]: as the <s>shared pointer</s> shared_ptr <s>they point to</s> reference may change

File: test/framework/integration_framework/fake_peer/fake_peer.hpp [hunk 7]
```diff
@@ -10,169 +10,229 @@
...
+       * @param listen_ip - IP on which this fake peer should listen
+       * @param internal_port - the port for internal commulications
+       * @param key - the keypair of this peer
+       * @param real_peer - the main tested peer managed by ITF
+       * @param common_objects_factory - common_objects_factory
+       * @param transaction_factory - transaction_factory
+       * @param batch_parser - batch_parser
+       * @param transaction_batch_factory - transaction_batch_factory
+       * @param gree_all_proposals - whether this peer should agree all
+       * proposals
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/fake_peer.hpp]: Sweet yeti of the Serengeti!
That a tremendous number of ctor args. Probably not an issue of the PR, but that tech debt should be considered at least in future

File: test/framework/integration_framework/fake_peer/proposal_storage.cpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  namespace fake_peer {
+
+    ProposalStorage::ProposalStorage()
+        : default_provider_([](auto &) { return boost::none; }) {}
+
+    OrderingProposalRequestResult ProposalStorage::getProposal(
+        const Round &round) const {
+      auto it = proposals_map_.find(round);
+      if (it != proposals_map_.end()) {
+        if (it->second) {
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/proposal_storage.cpp]: might be a bit clearer:
`return it->second ? OrderingProposalRequestResult{*it->second}: boost::none;`

File: test/framework/integration_framework/fake_peer/types.hpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  }    // namespace consensus
+  namespace ordering {
+    namespace transport {
+      class OnDemandOsServerGrpc;
+    }
+    class OrderingGateTransportGrpc;
+    class OrderingServiceTransportGrpc;
+  }  // namespace ordering
+  class MstState;
+}  // namespace iroha
```
Comment (l4l) [inline test/framework/integration_framework/fake_peer/types.hpp:63]: I guess these fds should be in core, rather in tests

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 10]
```diff
@@ -634,4 +675,9 @@ namespace integration_framework {
       boost::filesystem::remove_all(iroha_instance_->block_store_dir_);
     }
   }
+
+  const std::shared_ptr<IrohaInstance>
+      &IntegrationTestFramework::getIrohaInstance() const {
```
Comment (l4l) [inline test/framework/integration_framework/integration_test_framework.cpp]: `const shared_ptr<..>`? :\ Maybe just `const IrohaInstance&`?

Also I'm not really sure that this component should be publicly available

File: test/integration/acceptance/basic_mst_state_propagation.cpp [hunk 11]
```diff
@@ -1,111 +0,0 @@
-/**
```
Comment (l4l) [inline test/integration/acceptance/basic_mst_state_propagation.cpp:1]: Why is that removed?

File: test/framework/integration_framework/fake_peer/block_storage.cpp (hunk 3 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/block_storage.cpp:85]: Comrade l4l, by using `at` I strengthen the immutability of the map!

File: test/framework/integration_framework/fake_peer/proposal_storage.cpp (hunk 8 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/proposal_storage.cpp]: the ternary operator wants both alternatives of  the same type, and accomplishing this makes the code way uglier than the current version, which is at least easily readable.

File: test/integration/acceptance/basic_mst_state_propagation.cpp (hunk 11 above)
Comment (MBoldyrev) [reply test/integration/acceptance/basic_mst_state_propagation.cpp:1]: it is replaced with `test/integration/acceptance/fake_peer_example_test.cpp`

File: shared_model/interfaces/iroha_internal/transaction_batch.hpp [hunk 12]
```diff
@@ -53,6 +53,9 @@ namespace shared_model {
           size_t number_of_tx,
           const shared_model::crypto::Signed &signed_blob,
           const shared_model::crypto::PublicKey &public_key) = 0;
+
+      /// Pretty print the batch contents.
```
Comment (igor-egorov) [inline shared_model/interfaces/iroha_internal/transaction_batch.hpp:57]: minor
Seems that just "batch contents" (without "the") will be ok.

File: test/framework/integration_framework/fake_peer/behaviour/honest.cpp [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+      getFakePeer().voteForTheSame(message);
+    }
+
+    std::string HonestBehaviour::getName() {
+      return "honest behaviour";
+    }
+
+    LoaderBlockRequestResult HonestBehaviour::processLoaderBlockRequest(
+        LoaderBlockRequest request) {
+      const auto block_storage = getFakePeer().getBlockStorage();
```
Comment (Akvinikym) [inline test/framework/integration_framework/fake_peer/behaviour/honest.cpp]: Why not `const auto &` here and below?

File: test/framework/integration_framework/integration_test_framework.hpp [hunk 14]
```diff
@@ -251,6 +277,32 @@ namespace integration_framework {
      */
     IntegrationTestFramework &sendQuery(const shared_model::proto::Query &qry);
 
+    /// Send proposal to this peer' s ordering service.
```
Comment (Akvinikym) [inline test/framework/integration_framework/integration_test_framework.hpp]:
```suggestion
    /// Send proposal to this peer's ordering service.
```

File: test/framework/integration_framework/fake_peer/network/mst_network_notifier.hpp [hunk 15]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FAKE_PEER_MST_NETWORK_NOTIFIER_HPP_
+#define FAKE_PEER_MST_NETWORK_NOTIFIER_HPP_
+
+#include <rxcpp/rx.hpp>
+
```
Comment (igor-egorov) [inline test/framework/integration_framework/fake_peer/network/mst_network_notifier.hpp]: redundant line break

File: test/framework/integration_framework/fake_peer/network/ordering_service_network_notifier.hpp [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FAKE_PEER_OS_NETWORK_NOTIFIER_HPP_
+#define FAKE_PEER_OS_NETWORK_NOTIFIER_HPP_
+
+#include "network/ordering_service_transport.hpp"
+
+#include <rxcpp/rx.hpp>
+
```
Comment (igor-egorov) [inline test/framework/integration_framework/fake_peer/network/ordering_service_network_notifier.hpp]: redundant line break

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace integration_framework {
+  namespace fake_peer {
+
+    Behaviour::~Behaviour() {
+      absolve();
+    }
+
+    void Behaviour::adopt(const std::shared_ptr<FakePeer> &fake_peer) {
+      // This code feels like part of constructor, but the use of `this'
```
Comment (igor-egorov) [inline test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp:19]: please use the same type of enclosing quotes around `this`

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+      // Hint: such calls would precede the derived class construction.
+      fake_peer_wptr_ = fake_peer;
+      log_ = logger::log(getName() + " of fake peer at "
+                         + fake_peer->getAddress());
+      // subscribe for all messages
+      subscriptions_.emplace_back(
+          getFakePeer().getMstStatesObservable().subscribe(
+              [this](const auto &message) {
+                this->processMstMessage(message);
+              }));
```
Comment (igor-egorov) [inline test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp:53]: can we avoid intermediary lambdas and pass class methods like `processMstMethods` directly to `subscribe` call?

something like
```
getFakePeer().getMstStatesObservable().subscribe(processMstMessage);
```

File: test/framework/integration_framework/fake_peer/behaviour/honest.cpp (hunk 13 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/behaviour/honest.cpp]: here because the method signature is
```c++
      boost::optional<const BlockStorage &> getBlockStorage() const;
```
neither below I can see where to make this change.

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.hpp (hunk 1 above)
Comment (igor-egorov) [reply test/framework/integration_framework/fake_peer/behaviour/behaviour.hpp]: logging method calls also is an option. Please consider

File: test/framework/integration_framework/fake_peer/behaviour/honest.cpp (hunk 13 above)
Comment (igor-egorov) [reply test/framework/integration_framework/fake_peer/behaviour/honest.cpp]: Probably Pavel meant the line 42
```
const auto block_storage = getFakePeer().getBlockStorage();
```

File: test/framework/integration_framework/fake_peer/block_storage.cpp (hunk 2 above)
Comment (igor-egorov) [reply test/framework/integration_framework/fake_peer/block_storage.cpp:20]: why not just `operator[]` ?


```
result = end() == find(key)
map[key] = val
return result
```

File: test/framework/integration_framework/fake_peer/fake_peer.cpp [hunk 19]
```diff
@@ -119,6 +130,7 @@ namespace integration_framework {
 
     FakePeer &FakePeer::setBlockStorage(
         const std::shared_ptr<BlockStorage> &block_storage) {
+      ensureInitialized();
```
Comment (igor-egorov) [inline test/framework/integration_framework/fake_peer/fake_peer.cpp:173]: Why we can't use `initialize()` here instead `ensureInitialized()`?

Can we put such check into `initialize()`?
```diff
    FakePeer &FakePeer::initialize() {
-      BOOST_VERIFY_MSG(!initialized_, "Already initialized!"); 
+      if (not initialized_) {
        // here comes the initialization of members requiring shared_from_this()
        synchronizer_transport_ =
            std::make_shared<LoaderGrpc>(shared_from_this());

        initialized_ = true;
+      }
      return *this;
    }
```

File: test/framework/integration_framework/fake_peer/fake_peer.cpp [hunk 20]
```diff
@@ -106,8 +106,19 @@ namespace integration_framework {
           + getAddress() + ")");
     }
 
+    FakePeer &FakePeer::initialize() {
+      BOOST_VERIFY_MSG(!initialized_, "Already initialized!");
```
Comment (igor-egorov) [inline test/framework/integration_framework/fake_peer/fake_peer.cpp]: our practice is using of `not,and,or` instead of `!,&&,||`

File: test/framework/integration_framework/integration_test_framework.cpp [hunk 21]
```diff
@@ -534,11 +535,43 @@ namespace integration_framework {
...
 
+  IntegrationTestFramework &IntegrationTestFramework::sendBatches(
+      const iroha::consensus::Round &round,
+      const std::vector<TransactionBatchSPtr> &batches) {
+    auto on_demand_os_transport =
+        iroha::ordering::transport::OnDemandOsClientGrpcFactory(
+            async_call_,
+            [] { return std::chrono::system_clock::now(); },
+            std::chrono::milliseconds(0)  // the proposal waiting
+                                          // timeout does not apply here
```
Comment (igor-egorov) [inline test/framework/integration_framework/integration_test_framework.cpp]: please change the comment.

unfortunately, the comment brings more confusion than clarification.
Please add the reason, **why** timeout does not apply here.

without the clarification, the comment just states the fact without explanation

File: test/integration/acceptance/fake_peer_example_test.cpp [hunk 22]
```diff
@@ -274,14 +282,17 @@ TEST_F(FakePeerExampleFixture,
...
+ *
  * \attention this code is nothing more but an example of Fake Peer usage
  *
  * @given a network of two iroha peers
  * @when a valid command is sent to one
  * @then it must propagate either a proposal or a batch
  */
 TEST_F(FakePeerExampleFixture,
-       OrderingMessagePropagationAfterValidCommandReceived) {
+       DISABLED_OrderingMessagePropagationAfterValidCommandReceived) {
```
Comment (igor-egorov) [inline test/integration/acceptance/fake_peer_example_test.cpp]: Please create a task with explanation why the test is disabled and add a todo here.

File: test/framework/integration_framework/fake_peer/behaviour/honest.cpp (hunk 13 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/behaviour/honest.cpp]: Yes, now I got it, we can assign an rvalue to a const lvalue reference.

Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/behaviour/honest.cpp]: Yes, now I got it, we can assign an rvalue to a const lvalue reference.

File: test/framework/integration_framework/fake_peer/fake_peer.cpp (hunk 19 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/fake_peer.cpp:173]: I would like to force initialization at creation time before the object starts being used. That is why there is an assertion.

File: irohad/ordering/impl/ordering_gate_transport_grpc.cpp [hunk 23]
```diff
@@ -17,7 +17,7 @@ grpc::Status OrderingGateTransportGrpc::onProposal(
     ::grpc::ServerContext *context,
     const iroha::protocol::Proposal *request,
     ::google::protobuf::Empty *response) {
-  async_call_->log_->info("receive proposal");
+  async_call_->log_->info("received a proposal");
```
Comment (igor-egorov) [inline irohad/ordering/impl/ordering_gate_transport_grpc.cpp]: test single comment

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp (hunk 18 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp:53]: Seems we cannot as the `this` pointer has to be bound somehow to the callable, and a lambda is a cheap way to do that.

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp (hunk 17 above)
Comment (MBoldyrev) [reply test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp:19]: This type of quotes is widely used in program documentation and I personally like it.

File: test/integration/acceptance/fake_peer_example_test.cpp (hunk 22 above)
Comment (MBoldyrev) [reply test/integration/acceptance/fake_peer_example_test.cpp]: I found no task about removing the old ordering service code, so we decided the comment above the test code is enough.

File: test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp (hunk 17 above)
Comment (igor-egorov) [reply test/framework/integration_framework/fake_peer/behaviour/behaviour.cpp:19]: D:

**Comment by MBoldyrev** [comment]
Need to fix some multithreading concurrency issues prior to merging this.
